highlight-range.js

```diff
@@ -1,19 +1,48 @@
 var highlightRange = (function () {
-// Wrap each text node in a given DOM Range with a <span class=[highLightClass]>.
+// Wrap each text node in a given DOM Range with an inline node , defaulting to <span class=[highLightClass]>.
 // Breaks start and/or end node if needed.
 // Returns a function that cleans up the created highlight (not a perfect undo: split text nodes are not merged again).
 //
 // Parameters:
 // - rangeObject: a Range whose start and end containers are text nodes.
-// - highlightClass: the CSS class the text pieces in the range should get, defaults to 'highlighted-range'.
-function highlightRange(rangeObject, highlightClass) {
+// - highlightTemplate
+//     string: the CSS class the text pieces in the range should get, defaults to 'highlighted-range'
+//     HTMLElement: an element that will be cloned for each highlight
+//     function: a function that creates a wrapper node. It's passed the Node that's being
+//               highligted (but is free to ignore it). This can be used to apply properties 
+//               to a logical highlight, even if it's spread over many elements, for example
+//               to store an annotation identifier to be able to query for all highlight
+//               spans for the same annotation:
+//               
+//                 function renderAnnotation(annotation, range) {
+//                   highlightRange(range, function() {
+//                     var span = document.createElement('span');
+//                     span.classList.add('annotation');
+//                     span.dataset.annotationId = annotation.id;
+//                     return span;
+//                   });
+//                 }
+function highlightRange(rangeObject, highlightTemplate) {
     // Ignore range if empty.
     if (rangeObject.collapsed) {
         return;
     }
 
-    if (typeof highlightClass == 'undefined') {
-        highlightClass = 'highlighted-range';
+    var highlightCallback = highlightTemplate || defaultHighlight;
+    if ('string' === typeof highlightTemplate) {
+      highlightCallback = function _highlightTemplateClassName() {
+        var span = document.createElement('span');
+        span.classList.add(highlightTemplate);
+        return span;
+      };
```

Comment:
This function is nearly the same as `defaultHighlight`, I would make `defaultHighlight` accept a class name as second argument so it can be used here: `highlightCallback = node => defaultHighlight(node, highlightTemplate)`

Reply:
Yes, that’s cleaner.